[LLP] Memory safe buffered ledger api streams [DPP-1020]

tudor-da opened the pull request:
Currently, a Ledger API stream subscription can keep alive a full in-memory `EventsBuffer` fan-out buffer size  (e.g. consider a client subscribes that to `TransactionsService.GetTransactions` and stops consuming). Worst-case memory footprint is `O(buffer_size*client_subscriptions)`.

This PR reduces the worst-case memory footprint by chunking the fetches from the `EventsBuffer` with a configurable chunk size. The worst case memory footprint is `O(buffered-streams-page-size*client_subscriptions)`.

There are two commits in this PR:
* The first commit is only for proving the hypothesis - it contains a test that causes an OOM when run - it exercises the worst-case scenario on the `BufferedTransactionsReader`. 
* The second commit contains the implementation. Additionally, one can check the test created in the first commit to observe that it passes. Some guiding details about the implementation:
  * Limited size chunks are formed in the `EventsBuffer` when the resulting slice exceeds the maximum chunk size. Fetching multiple chunks per dispatcher fetch window is done recursively (see `BufferedTransactionsReader`).
  * Transaction chunks must contain already filtered transactions. For this reason, `TransactionLogUpdateConversions` is refactored for separating the transaction filter (performed before chunk forming in `EventsBuffer.slice`) to the API domain transaction conversion (performed in `BufferedTransactionsReader`) since filtering needs to happen before chunking in `BufferEvents.slice`
  *  `buffered-streams-page-size` is the CLI config parameter configuring the max chunk size.

**NOTE**: `MemorySafeBufferedTransactionsReaderSpec` proves the hypothesis of this implementation. It will NOT be merged due to the long run-times (> 20-30 seconds) and high memory pressure on the CI. I will remove it before merging.

<!--
# Pull Request Checklist

- Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- Include appropriate tests
- Set a descriptive title and thorough description
- Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- Normal production system change, include purpose of change in description
- If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.
-->

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesReader.scala:
```diff
@@ -34,9 +34,9 @@ object TransactionLogUpdatesReader {
-          treeEventWitnesses = raw.treeEventWitnesses,
-          flatEventWitnesses = raw.flatEventWitnesses,
-          submitters = raw.submitters,
+          treeEventWitnesses = raw.treeEventWitnesses.map(Ref.Party.assertFromString),
+          flatEventWitnesses = raw.flatEventWitnesses.map(Ref.Party.assertFromString),
+          submitters = raw.submitters.map(Ref.Party.assertFromString),
```
Changes not imposed by the scope of this PR, but I've done it for having a homogeneous domain representation of types in the `TransactionLogUpdate` DTOs.

tudor-da commented on ledger/participant-integration-api/src/test/suite/scala/platform/store/appendonlydao/events/MemorySafeBufferedTransactionsReaderSpec.scala:
```diff
@@ -0,0 +1,150 @@
+
+import scala.concurrent.Future
+
+class MemorySafeBufferedTransactionsReaderSpec
```
This test takes a long time (20-30 seconds) and exerts high memory pressure. It is here only for proving the hypothesis and resolution

- [ ] Remove before merging

skisel-da commented on ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:
```diff
@@ -21,141 +21,231 @@ import scala.concurrent.duration.DurationInt
-        }
-        Succeeded
+            Await.result(
+              // Simulate different thread accesses for push/slice
```
why `different thread` if it is a fixed single-thread pool?

skisel-da replied:
ah I see, never mind

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+          val bufferSlice = vectorSnapshot.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
+
+          if (bufferStartInclusiveIdx == 0) {
+            // The buffer is potentially a suffix of the fetched window
```
Maybe `requested` rather than `fetched` ?

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+  private def searchByOffset(buffer: Vector[(Offset, ENTRY)], offset: Offset) =
+    buffer.searchBy(offset, _._1) match {
+      case InsertionPoint(insertionPoint) => insertionPoint
+      case Found(foundIndex) => foundIndex + 1
```
When we are using this with the before index do we still want to add 1 ?

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+              Suffix(endInclusive, Vector.empty, None)
+            } else {
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice.tail, filter)
+              // We "waste" the first element of the slice in order to provide
```
I guess we could be smarter about this is we considered if the `buffer.searchBy` returned `Found`.  Only possibly worth considering from a latency perspective in the case of a single transaction being played in ?

simonmaxen-da replied:
What I was trying to say was that if the full range was found in the buffer, starting at position 0, then maybe an `Inclusive` could be returned instead of a `Suffix` (so no need for `endInclusive`) ?

tudor-da replied:
This is the current behavior. if the index is found in the buffer at `Found(0)`, `bufferStartInclusiveIdx` is `1` which takes the `Inclusive(..)` branch

tudor-da replied:
I understand it was not trivial to follow what's going on. Tried a refactoring for better visibility: bfba6d372cf7ff2cf3757c6108a00c6719ea0666

simonmaxen-da replied:
The changes make it much clearer what is going on, thanks.

simonmaxen-da commented on ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:
```diff
@@ -21,141 +21,231 @@ import scala.concurrent.duration.DurationInt
-        buffer.slice(0, LastOffset) shouldBe Prefix(BufferElements.drop(2))
-        buffer.push(21, 42)
-        buffer.slice(0, 21) shouldBe Prefix(BufferElements.drop(3) :+ 21 -> 42)
+        buffer.slice(BeginOffset, LastOffset, IdentityFilter) shouldBe Suffix(
```
If we were to use `offset1` as the begin offset then bufferedStartExclusive may be `offset3` (due to double adjustment).

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+    (filteredEntries, Option.when(it.hasNext)(lastScannedOffset))
+  }
+
+  private def searchByOffset(buffer: Vector[(Offset, ENTRY)], offset: Offset) =
```
As we are adding 1 maybe a better name for this would be `indexAfterOffset`

tudor-da replied:
Good idea 👍

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -145,87 +160,132 @@ private[platform] object BufferedTransactionsReader {
-          resolvedFromBufferCounter.inc()
-          offset -> apiResponseCtor(Seq(tx))
+        .buffered(128)(inStreamBufferLength)
+        .mapAsync(1) { case (offset, payload) =>
```
What was the thinking behind dropping from 32 to 1 ?

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -145,87 +160,132 @@ private[platform] object BufferedTransactionsReader {
+              case _ => None
+            },
+          ) match {
+            case BufferSlice.Suffix(headOffset, tail, continue) =>
```
Not exactly this, but made it less verbose and more readable: bfba6d372cf7ff2cf3757c6108a00c6719ea0666

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -84,9 +100,9 @@ private[events] object TransactionLogUpdatesConversions {
 
     private val FlatTransactionPredicate =
```
bfba6d372cf7ff2cf3757c6108a00c6719ea0666

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -182,43 +194,40 @@ private[events] object TransactionLogUpdatesConversions {
-                  // childEventIds need to be returned in the event order in the original transaction.
-                  // Unfortunately, we did not store them ordered in the past so we have to sort it to recover this order.
-                  // The order is determined by the order of the events, which follows the event order of the original transaction.
-                  .sortChildEventIdsBy(visibleOrder)
```
Why do you think you can remove this? Also changes above: you seem like rolling back my changes to fix child event id sot order. Is this a copy paste issue? Or will this be done someplace else?

tudor-da replied:
No, this was not intentional, I suspect this might be due to a rebase. However, why didn't the newly introduced test fail?

tudor-da replied:
Which makes me realize: We don't actually need this fix here. The in-memory fan-out buffers cannot be populated with data ingested by an older process (which might have run the indexer that was not sorting these).

nmarton-da replied:
yes, this "double fixing" has the same issues with the normal persistent TransactionReader path as well. Conformance tests only uncover problems if the the indexer fix is manually reverted.
On your offline comment (whether this fix for the IMFO is needed at all, since IMFO only targets new data, which is stored already correctly with the indexer fix): you are generally right, but there could be corner cases with split-participant deployment, so for consistency on main, I would only remove the fix for this part, if split participant feature is removed.

nmarton-da replied:
after IMFO is fed from indexer directly (not from DB) this can be removed. (TODO)

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -10,66 +10,82 @@ import com.daml.ledger.api.v1.transaction.{
+      val nonTransient = removeTransient(transaction.events)
+      val requestingParties = filter.keySet
+      Future
+        .traverse(nonTransient)(toFlatEvent(_, requestingParties, verbose, lfValueTranslation))
```
just by doing the Future traverse here we will sync execute all toFlatEvent future factory functions, which unfortunately boil down to sync Engine call if verbose mode is on.
I advise caution executing lfValueTranslation code therefore, and wrapping the Future.traverse in a Future. An example for that you can see in TransactionReader.getActiveContracts:
```
          future = Future(
            Future.traverse(rawResult)(
              deserializeEntry(verbose)
            )
          ).flatMap(identity),
```
In the acs case it is more important because there deserialization logic is running async, bu since the engine call in this example I think it is still a good idea for being careful.

(we have pending tasks to fix the root cause here, but it was not scheduled/implemented yet)

nmarton-da replied:
until the toFlatEvent future factories are sync, they are also blocking in traverse, so as non-intuitively: this wont have non-limited parallelism (all entries executing in parallel) async, but rather sequential blocking sync call (upto the async part in the toFlatEvent method).
With wrapping this to a Future we would have that sequential (parallelism 1) async call.

tudor-da replied:
Good remarks. I'm leaving this change to the subsequent PR that is meant to address exactly these performance concerns (https://github.com/digital-asset/daml/pull/13880) in order to reduce the impact of this one since it was already validated from a performance POV.

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -182,43 +194,40 @@ private[events] object TransactionLogUpdatesConversions {
+      transaction =>
         Future
-          .traverse(filteredForVisibility)(
+          .traverse(transaction.events)(
```
Same here as I commented on the same function for flat above in this file (wrapping the traverse in a Future)

tudor-da replied:
Thanks, good remark. Will update in the subsequent performance improvement PR.

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
+      else
+        Source
+          .fromIterator(() => bufferSlice.iterator)
+          .mapAsync(1) { case (offset, payload) =>
```
A general remark what we discussed offline: I am slightly concerned whether we will know if this transformation is the bottleneck for stream delivery. As implementing the ACS 2 phase retrieval, I was making tests without verbose mode, an d with parallelism 2 it was never a bottleneck, but this might look very differently here.

In the ACS implementation in TransactionReader this is coming from a static config (set to two). I am wondering whether we should do here the same.

(of course this only helps in a logical way if the toApiTx is async, see my corresponding comments in the conversions file, next to the implementation. There I also elaborate why it cannot always be expected the implementation to run in parallel at all for one such call)

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
-      }),
-    )
+      }
+      .buffered(outputStreamBufferSize)(bufferReaderMetrics.bufferSize)
```
As discussed offline we need to move the output buffer outside of Dispatcher stream consumption (as also discussed, in a different PR)

tudor-da replied:
As discussed offline, preparing a subsequent PR for fixing this generally in our streams assembly logic.

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,28 +89,40 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+            case InsertionPoint(0) if bufferSlice.isEmpty =>
+              Suffix(endInclusive, Vector.empty, None)
+            case InsertionPoint(0) =>
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice.tail, filter)
```
As discussed offline we could have here the chunk calculated from backwards to struck a trade-off between more versatile stream-catch-up and having stream catch-up issues depend on the max buffer size, versus simpler logic but less versatile stream catch up, which depends on the max chink size.

tudor-da replied:
Done

nmarton-da replied:
:raised_hands:

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
+          }
+        case _ => None
+      }
+      .flatMapConcat(identity)
```
kudos for figuring out non-nested (and performant) recursive streams :clap:

nmarton-da reviewed (changes requested):
I block the PR because the suspicion it reverts important changes accidentally.
Otherwise the direction is solid! I am very happy to see the in-memory fan-out moving to the next level!
Great work!

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -3,44 +3,46 @@
+    isRangeEndMarker: ENTRY => Boolean,
+    maxBufferedChunkSize: Int,
 ) {
-  @volatile private var _bufferStateRef = BufferStateRef[O, E]()
```
(in the type)

tudor-da replied:
Done

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/interfaces/TransactionLogUpdate.scala:
```diff
@@ -99,11 +98,11 @@ object TransactionLogUpdate {
-      treeEventWitnesses: Set[String],
-      flatEventWitnesses: Set[String],
-      submitters: Set[String],
+      treeEventWitnesses: Set[Party],
```
I added my own GH comment at the inception of this PR motivating this.

nmarton-da replied:
Not sure what you are referring to. On any rate my comment is general in that sense: if it is not possible to extract it is fine, if it would be possible, next time I would consider to extract that. Quality of review can be improved if meaningful changes do not collide with pure refactorings.

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -131,6 +145,38 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+    val it = bufferSlice.iterator
+    val resultBuilder = Vector.newBuilder[(Offset, FILTER_RESULT)]
+
+    while (it.hasNext && takenFiltered < maxBufferedChunkSize) {
```
It was a conscious decision for one reason: checking whether we need to offer a new chunk (see line 177 with it.hasNext). I'd love making this functional, but couldn't come up with an efficient solution due to:
* This function should compute a filtered collection of size `maxBufferedChunkSize`
* It should also return a `startExclusive` needed for the next chunk processing IF the underlying window has more elements to-be-scanned.

tudor-da replied:
Improved somehow the implementation by making it more concise and removing the mutable vars. Although it's shorter now, I am now concerned a bit about how obvious is to everyone why the last `.hasNext` correctly flags that there needs to be another fetched chunk. Let me know what do you think

nmarton-da replied:
I like it a lot :clap: 
Re-using iterators is sometimes tricky, and generally discouraged in Scala, so I would suggest using splitAt, where you can covert _1 to Vector, and you can call the hasNext on _2

tudor-da replied:
If you use splitAt instead of take and call hasnext on _2, you are potentially evaluating a lot of filters (redundant work) until you reach a next Some(_) that yields true.

nmarton-da replied:
Oh yeah, that's right.
It makes me wonder whether we need this option here at all (there is already a signalling stream end for the unfold one level up, can we just make that smarter, and make this simpler)?
Similary to the lastFilteredChunk: there we could just do always the head/tail split for chunkSize+1, which would be even more efficient, since it would make the range for DB query shorter -- and similarly simplifying this function.

nmarton-da replied:
As discussed offline improvements here will come later (maybe a TODO)

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -3,44 +3,46 @@
     metrics: Metrics,
     bufferQualifier: String,
-    isRangeEndMarker: E => Boolean,
+    isRangeEndMarker: ENTRY => Boolean,
```
You are right, we can drop them after we implement batching with the fusing of the Indexer to the Ledger API.

nmarton-da replied:
I see so you are planning to do this later?

tudor-da replied:
Yes.

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +139,136 @@ private[events] class BufferedTransactionsReader(
+      .buffered(outputStreamBufferSize)(bufferReaderMetrics.bufferSize)
+  }
+
+  // TODO LLP: Consider merging with InstrumentedSource.bufferedSource
```
In a clean-up stage of this epic as this impacts non-related code paths and might stir up discussions on style (which should not block the progress of this implementation).

nmarton-da replied:
Sure :+1: 
An alternative would have been to have a small PR doing that which would precede this PR.

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -10,66 +10,82 @@ import com.daml.ledger.api.v1.transaction.{
 import com.daml.platform.participant.util.LfEngineToApi
-import com.daml.platform.store.interfaces.TransactionLogUpdate
 import com.daml.platform.store.interfaces.TransactionLogUpdate.{CreatedEvent, ExercisedEvent}
+import com.daml.platform.store.interfaces.{TransactionLogUpdate => TxUpdate}
```
This is just a shortcut in this context where everything is about `TransactionLogUpdate`s. Removing the alias since you it bothers you

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +139,136 @@ private[events] class BufferedTransactionsReader(
-          vs.foldLeft(acc) { case (a, v) =>
-            a + (v -> (a.getOrElse(v, Set.newBuilder) += k))
-          }
+      .foldLeft(Map.empty[Identifier, mutable.Builder[Party, Set[Party]]]) { case (acc, (k, vs)) =>
```
I think you are right. Simplifying

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -10,66 +10,82 @@ import com.daml.ledger.api.v1.transaction.{
 import scala.concurrent.{ExecutionContext, Future}
 
 private[events] object TransactionLogUpdatesConversions {
+  type Filter = TxUpdate.Transaction => Option[TxUpdate.Transaction]
```
Hmm, it's less verbose when referenced in other places. But fine by me to de-alias.

nmarton-da replied:
Not sure what you are referring to. On any rate my comment is general in that sense: if it is not possible to extract it is fine, if it would be possible, next time I would consider to extract that. Quality of review can be improved if meaningful changes do not collide with pure refactorings.

nmarton-da replied:
I understand the need for more complex types, I am merely raising this is not so complex. The tradeoff is the mental mapping what the author forces on the reviewer.

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala:
```diff
@@ -10,66 +10,82 @@ import com.daml.ledger.api.v1.transaction.{
+    ): Filter =
+      transaction => {
+        val transactionEvents = transaction.events
+        val flatTransactionEvents = transactionEvents.collect {
```
Done

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +139,136 @@ private[events] class BufferedTransactionsReader(
 
   private def invertMapping(partiesTemplates: Map[Party, Set[Identifier]]) =
```
Sure, done.

tudor-da commented:
Thanks for the in-depth review @nmarton-da . All your concerns/suggestions are addressed. I have left un-addressed some comments which regard performance, since there is already a subsequent [PR](https://github.com/digital-asset/daml/pull/13880) to bundle these in after this one is merged (since the suggestions are touching also the existing code, not only what is introduced in this PR).

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala:
```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
+      if (bufferSlice.isEmpty) Source.empty
+      else
+        Source
+          .fromIterator(() => bufferSlice.iterator)
```
None. I was not aware of this simple constructor

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -193,4 +208,53 @@ private[platform] object EventsBuffer {
+      case Found(foundIndex) => foundIndex + 1
+    }
+
+  private[cache] def filterAndChunkSlice[ENTRY, FILTER_RESULT](
```
general question about unit test coverage: do you later plan to increase unit test coverage for these nicely extracted functions?

tudor-da replied:
They are unit tested: https://github.com/digital-asset/daml/pull/13852/files#diff-2155de267cfec3ba1419d81f983111ffac90fd7a74d72a5316423ecfec89f502R289-R357

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -193,4 +208,53 @@ private[platform] object EventsBuffer {
+  ): LastBufferChunkSuffix[(Offset, FILTER_RESULT)] = {
+    val reversedSliceIterator = bufferSlice.view.reverseIterator
+
+    val lastChunk =
```
It would be very nice to make this code a little bit more DRY (the iterator transformations and the Option.when part look very similar)

tudor-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -193,4 +208,53 @@ private[platform] object EventsBuffer {
         }
       }
   }
+
```
Yeah, the time has come for that one. In a different PR ;)

nmarton-da commented on ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:
```diff
@@ -87,28 +89,40 @@ private[platform] final class EventsBuffer[O: Ordering, E](
+              // for back-filling the window elements that are before the buffer's bounds.
+              Suffix(bufferSlice.head._1, filteredSlice, continueFrom)
+            case InsertionPoint(_) | Found(_) =>
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice, filter)
```
as discussed this will come in a next PR for general perf improvements

nmarton-da reviewed (approved):
Looks very good!
Great work! :clap:
